### PR TITLE
feat: implement blind signature command

### PR DESCRIPTION
**Pull request by dj8yfo**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -75,21 +87,19 @@ pub fn get_version() -> Result<NEARLedgerAppVersion, NEARLedgerError> {
...
-                response.error_code()
             );
             // Ok means we successfully exchanged with the Ledger
             // but doesn't mean our request succeeded
             // we need to check it based on `response.retcode`
             if response.retcode() == RETURN_CODE_OK {
                 return Ok(response.data().to_vec());
             } else {
-                let error_string = response
-                    .error_code()
```
Comment (dj8yfo) [inline src/lib.rs:90]: `ledger-apdu-0.10.0` error code conversion table is probably from another app, its codes do not correspond to the ones in the `app-near` ledger repo

File: src/lib.rs [hunk 2]
```diff
@@ -229,11 +230,12 @@ pub fn get_public_key(
...
 /// ```
-/// near_crypto::Signature::from_parts(near_crypto::KeyType::ED25519, &signature)
-///     .expect("Signature is not expected to fail on deserialization")
+/// # let signature = [10; 64].to_vec();
+/// let signature = near_crypto::Signature::from_parts(near_crypto::KeyType::ED25519, &signature)
+///     .expect("Signature is not expected to fail on deserialization");
 /// ```
 pub fn sign_transaction(
-    unsigned_transaction_borsh_serializer: BorshSerializedUnsignedTransaction,
+    target: SignTarget,
```
Comment (frol) [inline src/lib.rs]: To be honest, from a crate user perspective, I would prefer to have a single `sign_transaction` method that would figure out when it should use blind signing or not, and maybe additionally have 3 lower-level methods:

* non-blind transaction signing (receive a serialized transaction)
* blind transaction signing (receive a serialized transaction)
* blind transaction hash signing (receive a hash)

**Comment by dj8yfo** [comment]
discovered a marginal bug when reiterating a test with signing a normal tx, fitting in buffer
[LedgerHQ/app-near#31](https://github.com/LedgerHQ/app-near/pull/31)

UPD: actually, the bug is not marginal, and this line affects the agreed upon flow with first attempting to sign normally, 
https://github.com/LedgerHQ/app-near/pull/31/files#diff-e8ced7bbc279c2661eb12f8ed3505734a06f5d267cc934852019629fe2f2601bR363 , as `app-near` was not resetting signing context on attempting to sign a too large transaction.

File: examples/blind_sign_transaction.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+use near_primitives::types::AccountId;
+
+use near_primitives::borsh::BorshSerialize;
+use near_primitives_core::hash::CryptoHash;
+use slip10::BIP32Path;
+
+fn too_long_tx() -> near_primitives::transaction::Transaction {
+    let ledger_pub_key_str = "7be181e38cf773f8432a1af401f83b39f1222bad5a167875abba1baa2de477c7";
+    let receiver_pub_key_str = "dc7e34eecec3096a4a661e10932834f801149c49dba9b93322f6d9de18047f9c";
+    let ledger_pub_key_bytes = hex::decode(ledger_pub_key_str).unwrap();
```
Comment (frol) [inline examples/blind_sign_transaction.rs]: Given that public/secret keys are encoded with base58 in NEAR, let's use base58 here as well to avoid confusion.

File: examples/blind_sign_transaction.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    let ledger_pub_key = near_crypto::PublicKey::ED25519(
+        near_crypto::ED25519PublicKey::try_from(ledger_pub_key_bytes.as_ref()).unwrap(),
+    );
+
+    let block_hash_str = "Cb3vKNiF3MUuVoqfjuEFCgSNPT79pbuVfXXd2RxDXc5E";
+
+    let block_hash_bytes = block_hash_str.from_base58().unwrap();
+    let mut block_hash: [u8; 32] = [0; 32];
+    block_hash.copy_from_slice(&block_hash_bytes[0..32]);
+    let block_hash = near_primitives::hash::CryptoHash(block_hash);
```
Comment (frol) [inline examples/blind_sign_transaction.rs]: Use `CryptoHash::from_str` or even `"...".parse::<CryptoHash>()`

File: examples/sign_transaction.rs [hunk 5]
```diff
@@ -35,18 +37,41 @@ fn main() -> Result<(), NEARLedgerError> {
...
-    log::info!("{:#?}", unsigned_transaction);
+pub fn verify_near(
+    bytes: &[u8],
+    pub_key: &PublicKey,
+    signature: &Signature,
+) -> Result<(), SignatureError> {
+    let hash = CryptoHash::hash_bytes(bytes);
+
+    pub_key.verify(&hash.as_ref(), &signature)
+}
```
Comment (frol) [inline examples/sign_transaction.rs]: It feels like the function body is actually simpler than the signature. I would just inline it

File: examples/blind_sign_transaction.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    let bytes = unsigned_transaction
+        .try_to_vec()
+        .expect("Transaction is not expected to fail on serialization");
+    log::info!("bytes len : {}", bytes.len());
+    let err = near_ledger::sign_transaction(bytes.clone(), hd_path.clone()).unwrap_err();
+
+    let hash = OnlyBlindSigning::hash_bytes(&bytes);
+    log::info!("{:<25} : {}", "hash (hex)", hex::encode(hash.0.as_ref()));
+    log::info!("{:<25} : {}", "hash", hash.0);
+    assert!(matches!(err, NEARLedgerError::BufferOverflow(err_hash) if err_hash == hash));
```
Comment (frol) [inline examples/blind_sign_transaction.rs]: This looks like an abuse of an example. It seems like you wanted to implement a test and example at the same time, but the downside now is that the example is hard to follow due to verbosity.

File: src/lib.rs [hunk 7]
```diff
@@ -169,28 +193,41 @@ pub fn get_public_key(
...
+    // will return an error if Ledger is not connected
+    let hidapi = match HidApi::new() {
+        Ok(hidapi) => hidapi,
+        Err(_err) => return Err(NEARLedgerError::DeviceNotFound),
+    };
+    match TransportNativeHID::new(&hidapi) {
+        Ok(transport) => Ok(transport),
+        // TODO: refactor this
+        // https://github.com/Zondax/ledger-rs/issues/65
+        Err(_err) => Err(NEARLedgerError::DeviceNotFound),
```
Comment (frol) [inline src/lib.rs]: Since the issue is resolved, can we address this TODO?

File: src/lib.rs [hunk 8]
```diff
@@ -27,8 +41,23 @@ pub enum NEARLedgerError {
...
+    /// Blind signature disabled in ledger's app settings
+    BlindSignatureDisabled,
     /// Error with transport
     LedgerHIDError(LedgerHIDError),
+    /// Transaction is too large to be signed
+    BufferOverflow(OnlyBlindSigning),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct OnlyBlindSigning(pub CryptoHash);
```
Comment (frol) [inline src/lib.rs]: I would use struct with a named field here to explicitly say what that CryptoHash is going to be.

File: src/lib.rs [hunk 9]
```diff
@@ -27,8 +41,23 @@ pub enum NEARLedgerError {
...
+
+impl OnlyBlindSigning {
+    pub fn hash_bytes(bytes: &[u8]) -> Self {
```
Comment (frol) [inline src/lib.rs]: I don't see any value in exposing this method to be part of the public API. Do you?

**Review by frol — CHANGES_REQUESTED** [review]
Nice! I have only left some nit comments, but overall it is in a pretty good state

File: src/lib.rs (hunk 7 above)
Comment (dj8yfo) [reply src/lib.rs]: addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/25b16bd49998b08d85aba6ce177ce453dc2c4948)

File: src/lib.rs (hunk 8 above)
Comment (dj8yfo) [reply src/lib.rs]: addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/a179ffa6791569cb599f630bab163bf1129f9c4c)

Comment (dj8yfo) [reply src/lib.rs]: addressed in [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/26d0f39f09c5d59f7d9a8c029041a795022d4c67)

File: src/lib.rs (hunk 9 above)
Comment (dj8yfo) [reply src/lib.rs]: addressed by [commit](https://github.com/khorolets/near-ledger-rs/pull/4/commits/245d912a5ae3c6dfff2af450334989f48a4e4fbd)

File: examples/sign_transaction.rs (hunk 5 above)
Comment (dj8yfo) [reply examples/sign_transaction.rs]: addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/6c7d11440a8c1f391e4d11aa726354f7885a03c5

File: examples/blind_sign_transaction.rs (hunk 3 above)
Comment (dj8yfo) [reply examples/blind_sign_transaction.rs]: the hex representation is used for computing `signer_id`/`receiver_id` `AccountId`-s
Attempting to do:
```rust
    let account_id = AccountId::from_str("9Lab2du3zZg5MrZPJCAd1fka7nHuP9Rgo643d8C8hGiJ").unwrap();
```
results in 
```bash
called `Result::unwrap()` on an `Err` value: ParseAccountError { kind: InvalidChar, char: Some((1, 'L')) }
```
and
```rust
    let account_id = AccountId::from_str("ed25519:9Lab2du3zZg5MrZPJCAd1fka7nHuP9Rgo643d8C8hGiJ").unwrap();
```
in 
```bash
called `Result::unwrap()` on an `Err` value: ParseAccountError { kind: InvalidChar, char: Some((7, ':')) }
```
I'll add a commit which doesn't use the hex `str` for computing the pub key for transaction, but instead uses the public key queried from ledger (this should've been so initially for 2 signature examples)

Comment (dj8yfo) [reply examples/blind_sign_transaction.rs]: addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/6b030733d641434b06982cb51862d2684c435a4e

File: examples/blind_sign_transaction.rs (hunk 4 above)
Comment (dj8yfo) [reply examples/blind_sign_transaction.rs]: addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/e80c33931f219b1363680dd2786ccfd0b07a6f4f

File: examples/blind_sign_transaction.rs (hunk 6 above)
Comment (dj8yfo) [reply examples/blind_sign_transaction.rs]: assert of signature verification was useful for judging about correctness with minimal env (without need to attempt to send transactions over network) in this https://github.com/LedgerHQ/app-near/pull/27#discussion_r1385223992 case and any other, possibly future, case. 
Setting up fully automated tests (running in ci, not requiring an actual device) is possible, but will probably be hard/cumbersome task.
Hopefully this comment is addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/1da9710f0d472d2dd91c232ce1fc941e01b94b29

File: src/lib.rs [hunk 10]
```diff
@@ -16,19 +20,53 @@ const CHUNK_SIZE: usize = 128; // Chunk size to be sent to Ledger
...
     /// Error with transport
     LedgerHIDError(LedgerHIDError),
+    /// Transaction is too large to be signed
+    BufferOverflow(OnlyBlindSigning),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct OnlyBlindSigning {
+    pub byte_array: CryptoHash,
+}
```
Comment (frol) [inline src/lib.rs]: How about turning it into?

```suggestion
    BufferOverflow { transaction_hash: CryptoHash },
}
```

given that we only pull in near-primitives-core for CryptoHash, we may even consider having our own `type CryptoHash = [u8; 32];`

**Review by frol — APPROVED** [review]
I have one last question/suggestion, but overall I believe it is already in a good shape, so I approve it.

File: src/lib.rs (hunk 10 above)
Comment (dj8yfo) [reply src/lib.rs]: addressed by https://github.com/khorolets/near-ledger-rs/pull/4/commits/7729ef8bc0782688d2604f54f68109cfd0eb1865

**Review by frol — APPROVED** [review]
